ADAL/src/workplacejoin/mac/ADClientCertAuthHandler.m

```diff
@@ -85,4 +90,152 @@ + (BOOL)handleChallenge:(NSURLAuthenticationChallenge *)challenge
     return YES;
 }
 
++ (SecIdentityRef)promptUserForIdentity:(NSArray *)issuers
+                                   host:(NSString *)host
+                          correlationId:(NSUUID *)correlationId
+{
+    NSDictionary *query =
+    @{
+      (id)kSecClass : (id)kSecClassIdentity,
+      (id)kSecMatchIssuers : issuers,
+      (id)kSecMatchLimit : (id)kSecMatchLimitAll,
+      };
+    
+    CFTypeRef result = NULL;
+    
+    OSStatus status = SecItemCopyMatching((CFDictionaryRef)query, &result);
+    if (status == errSecItemNotFound)
+    {
+        AD_LOG_INFO(@"No certificate found matching challenge", correlationId, nil);
+        return nil;
+    }
+    else if (status != errSecSuccess)
+    {
+        AD_LOG_ERROR(([NSString stringWithFormat:@"Failed to find identity matching issuers with %d error.", status]), status, correlationId, nil);
+        return nil;
+    }
+    
+    return [ADCertificateChooserHelper showCertSelectionSheet:(__bridge NSArray *)result host:host correlationId:correlationId];
+}
+
+
++ (BOOL)handleChallenge:(NSURLAuthenticationChallenge *)challenge
+                session:(NSURLSession *)session
+                   task:(NSURLSessionTask *)task
+               protocol:(ADURLProtocol *)protocol
+      completionHandler:(ChallengeCompletionHandler)completionHandler;
+{
+#pragma unused(session)
+#pragma unused(task)
+    
+    NSUUID *correlationId = protocol.context.correlationId;
+    NSString *host = challenge.protectionSpace.host;
+    AD_LOG_INFO_F(@"Attempting to handle client certificate challenge", correlationId, @"host: %@", host);
+    
+    // See if this is a challenge for the WPJ cert.
+    NSArray<NSData*> *distinguishedNames = challenge.protectionSpace.distinguishedNames;
+    if ([self isWPJChallenge:distinguishedNames])
+    {
+        return [self handleWPJChallenge:challenge protocol:protocol completionHandler:completionHandler];
+    }
+    
+    // Otherwise check if a preferred identity is set for this host
+    SecIdentityRef identity = SecIdentityCopyPreferred((CFStringRef)host, NULL, (CFArrayRef)distinguishedNames);
```

Comment:
Good catch. I haven't seen a case, but looking up RFC 5246 7.4.4 it does mention that the list can be empty, so that implies it is possible we could get either nil or an empty list here. I'll re-craft the keychain query to make sure it doesn't crash on a nil issuers list.